metadata-ingestion/src/datahub/ingestion/api/source.py

```diff
@@ -35,22 +35,22 @@ class SourceReport(Report):
     events_produced_per_sec: int = 0
     event_ids: List[str] = field(default_factory=LossyList)
 
-    warnings: Dict[str, List[str]] = field(default_factory=dict)
-    failures: Dict[str, List[str]] = field(default_factory=dict)
+    warnings: LossyDict[str, LossyList[str]] = field(default_factory=LossyDict)
+    failures: LossyDict[str, LossyList[str]] = field(default_factory=LossyDict)
 
     def report_workunit(self, wu: WorkUnit) -> None:
         self.events_produced += 1
         self.event_ids.append(wu.id)
 
     def report_warning(self, key: str, reason: str) -> None:
-        if key not in self.warnings:
```

Comment:
there is a tricky scenario that can happen:

lossy_dict.get(k) returns False
then you set
lossy_dict[k] = []   # initializing an empty list

later, secure in the knowledge that you have set that key, you issue:
lossy_dict[k].append("my message")

Now, this last call might fail, because the previous assignment is not guaranteed to find an empty slot and succeed... 

As a result you have to change how you work with lossy_dicts ..
I haven't come up with a good way to hide this complexity.